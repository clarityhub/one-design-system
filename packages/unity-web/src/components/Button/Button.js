import React from 'react';
import cx from 'classnames';
import { css } from '@emotion/core';
import styled from '@emotion/styled';
import { bool, func, oneOf, oneOfType, node, string } from 'prop-types';
import { variants } from '@clarityhub/unity-core/lib/typography';
import borders from '@clarityhub/unity-core/lib/borders';
import colors from '@clarityhub/unity-core/lib/colors';
import { lighten, darken, desaturate } from '@clarityhub/unity-core/lib/utilities/color';

import Loading from '../svgs/Loading';

const Button = styled.button(
	css`
        overflow: visible;
        display: inline-block;

        ${variants.button.string}

        box-shadow: ${colors.shadow.default};
        cursor: pointer;
        text-align: center;
        white-space: nowrap;
        vertical-align: middle;
        user-select: none;
        border: 1px solid transparent;
        padding: .5rem 1.25rem;
        font-size: 1rem;
        line-height: 1.5;
        border-radius: ${borders.borderRadius.rem}rem;
        position: relative;
        text-decoration: none;
        text-transform: uppercase;
        transition: color .15s ease-in-out,background-color .15s ease-in-out,border-color .15s ease-in-out,box-shadow .15s ease-in-out;

        &:focus,
        &:hover {
            text-decoration: none;
        }

        &:focus,
        &.focused {
            outline: 0;
            box-shadow: 0 0 0 .2rem ${colors.focused.default};
        }

        &:disabled,
        &.disabled {
            cursor: not-allowed;
            pointer-events: none;
        }

        &:not(:disabled):not(.disabled) {
            cursor: pointer;
        }

        svg {
            height: 1rem;
            width: 1rem;
        }

        --text: ${colors.primary.default};
        --text-active: ${colors.black.default};
        --bg: transparent;
        --bg-active: transparent;
        --bg-disabled: transparent;

        &.-default,
        &.-primary,
        &.-danger,
        &.-white {
            color: var(--text);
            background-color: var(--bg);
            border-color: var(--bg);

            &.-outline {
                border-color: var(--outline, var(--bg));
                color: var(--outline-text, var(--bg));
            }

            &.-text {
                color: var(--text-text, var(--bg));
            }

            &:hover,
            &.active {
                background-color: var(--bg-active);
                border-color: var(--bg-active);
                color: var(--text-active);

                &.-outline {
                    background-color: var(--bg) !important;
                    border: 1px solid var(--bg);
                    color: var(--outline-text-active, var(--text));
                }

                &.-text {
                    color: var(--text-text-active, var(--bg-active));
                }
            }

            &:disabled,
            &.disabled {
                background-color: var(--bg-disabled);
                border-color: var(--bg-disabled);
                color: var(--text);

                &.-outline {
                    border: 1px solid var(--outline-disabled, var(--outline-text-disabled, var(--bg-disabled)));
                    color: var(--outline-text-disabled, var(--bg-disabled));
                }

                &.-text {
                    color: var(--text-text-disabled, var(--bg-disabled));
                }
            }
        }

        &.-primary {
            --text: ${colors.white.default};
            --text-active: ${colors.white.default};
            --outline-disabled: ${colors.gray.default};
            --bg: ${colors.primary.default};
            --bg-active: ${desaturate(darken(colors.primary.default))};
            --bg-disabled: ${desaturate(lighten(colors.primary.default))};
        }

        &.-danger {
            --text: ${colors.white.default};
            --text-active: ${colors.white.default};
            --outline-disabled: ${colors.gray.default};
            --bg: ${colors.danger.default};
            --bg-active: ${desaturate(darken(colors.danger.default))};
            --bg-disabled: ${lighten(colors.danger.default, 0.3)};
        }

        &.-default {
            --text: ${colors.black.default};
            --text-active: ${colors.black.default};
            --outline: ${colors.black.default};
            --outline-text: ${colors.black.default};
            --outline-text-active: ${colors.black.default};
            --outline-text-disabled: ${colors.gray.default};
            --text-text: ${colors.black.default};
            --text-text-active: ${colors.black.default};
            --text-text-disabled: ${colors.gray.default};
            --bg: ${colors.white.default};
            --bg-active: ${colors.dove.default};
            --bg-disabled: ${colors.muted.default};
        }

        &.-white {
            --text: ${colors.primary.default};
            --text-active: ${colors.black.default};
            --outline: ${colors.white.default};
            --outline-text: ${colors.white.default};
            --outline-text-active: ${colors.primary.default};
            --outline-text-disabled: ${colors.gray.default};
            --text-text: ${colors.white.default};
            --text-text-active: ${colors.white.default};
            --text-text-disabled: ${colors.gray.default};
            --bg: ${colors.white.default};
            --bg-active: ${colors.dove.default};
            --bg-disabled: ${colors.muted.default};
        }

        &.-text {
            background-color: transparent !important;
            border: 1px solid transparent;
            box-shadow: none;

            &:hover,
            &.active {
                border-color: transparent;
            }

            &.disabled,
            &:disabled {
                border: 1px solid transparent;
            }
        }

        &.-outline {
            background-color: transparent !important;
            border: 1px solid inherit;
            box-shadow: none;
        }

        &.disabled,
        &:disabled {
            cursor: not-allowed;
        }
    `,
	({ size }) => {
		switch (size) {
		case 'small':
			return css`
                font-size: 0.9rem;
                padding: .2rem 1rem;
            `;
		case 'large':
			return css`
                font-size: 1.2rem;
                padding: 0.8rem 1.4rem;
            `;
		case 'default':
		default:
			return '';
		}
	},

	({ block }) => block && css`
        display: block;
        width: 100%;
    `,

	({ loading }) => loading && css`
        color: transparent !important;
        cursor: wait !important;

        &:hover {
            color: transparent !important;
        }
    `,
);

const IconWrapper = styled.span`
    font-size: inherit;
    margin-right: 0.3rem;
    vertical-align: middle;

    svg {
        height: 1rem;
        width: 1rem;

        ${({ size }) => {
		switch (size) {
		case 'small':
			return css`
                height: 0.9rem;
                width: 0.9rem;
            `;
		case 'large':
			return css`
                height: 1.2rem;
                width: 1.2rem;
            `;
		case 'default':
		default:
			return '';
		}
	}}
    }
`;

const LoadingWrapper = styled.div`
    position: absolute;
    left: 50%;
    transform: translate(-50%, -50%);
    top: 50%;
    align-content: center;
    justify-content: center;
    display: flex;

    svg {
        height: 1.4rem;
        width: 1.4rem;

        ${({ size }) => {
		switch (size) {
		case 'small':
			return css`
                height: 1.2rem;
                width: 1.2rem;
            `;
		case 'large':
			return css`
                height: 2.4rem;
                width: 2.4rem;
            `;
		case 'default':
		default:
			return '';
		}
	}}

    path {
        ${({ type, outline, text }) => {
		switch (type) {
		case 'primary':
			return css`
                fill: ${colors.white.default};

                ${outline && css`
                    fill: ${colors.primary.default};
                `}

                ${text && css`
                    fill: ${colors.primary.default};
                `}

                &:hover,
                &.active {
                    ${outline && css`
                        fill: ${colors.white.default};
                    `}

                    ${text && css`
                        fill: ${desaturate(darken(colors.primary.default))};
                    `}
                }

                &:disabled,
                &.disabled {
                    ${outline && css`
                        fill: ${desaturate(lighten(colors.primary.default))};
                    `}

                    ${text && css`
                        fill: ${desaturate(lighten(colors.primary.default))};
                    `}
                }
            `;
		case 'danger':
			return css`
                fill: ${colors.white.default};

                ${outline && css`
                    fill: ${colors.danger.default};
                `}

                ${text && css`
                    fill: ${colors.danger.default};
                `}

                &:hover,
                &.active {
                    ${outline && css`
                        fill: ${colors.white.default};
                    `}

                    ${text && css`
                        fill: ${desaturate(darken(colors.danger.default))};
                    `}
                }

                &:disabled,
                &.disabled {
                    ${outline && css`
                        fill: ${desaturate(lighten(colors.danger.default, 0.3))};
                    `}

                    ${text && css`
                        fill: ${desaturate(lighten(colors.danger.default, 0.3))};
                    `}
                }
            `;
		case 'default':
		default:
			return css`
                fill: ${colors.black.default};

                ${outline && css`
                    fill: ${colors.black.default};
                `}

                ${text && css`
                    fill: ${colors.black.default};
                `}

                &:hover,
                &.active {
                    ${outline && css`
                        fill: ${colors.black.default};
                    `}

                    ${text && css`
                        fill: ${desaturate(darken(colors.black.default))};
                    `}
                }

                &:disabled,
                &.disabled {
                    fill: ${darken(colors.gray.default, 0.1)};

                    ${outline && css`
                        fill: ${lighten(colors.gray.default, 0.01)};
                    `}

                    ${text && css`
                        fill: ${desaturate(lighten(colors.black.default))};
                    `}
                }
            `;
		}}
}
        }
    }
`;

const renderIcon = (Icon) => {
	if (typeof Icon === 'function') {
		return <Icon />;
	}

	return Icon;
};

const ButtonWrapper = ({
	className,
	type,
	outline,
	text,
	size,
	icon,
	children, 
	component,
	loading,
	buttonType = 'button',
	...rest
}) => {
	let Component = Button;

	if (component) {
		Component = Button.withComponent(component);
	}

	return (
		<Component
			className={cx(className, {
				[`-${type}`]: type,
				'-outline': outline,
				'-text': text,
			})}
			outline={outline}
			text={text}
			size={size}
			loading={loading}
			type={buttonType}
			{...rest}
		>
			{
				icon && (
					<IconWrapper size={size}>{renderIcon(icon)}</IconWrapper>
				)
			}
			{
				loading && (
					<LoadingWrapper
						type={type}
						outline={outline}
						text={text}
						size={size}
						{...rest}
					>
						<Loading />
					</LoadingWrapper>
				)
			}
			{children}
		</Component>
	);
};

ButtonWrapper.propTypes = {
	block: bool,
	buttonType: string,
	component: oneOfType([string, func]),
	icon: oneOfType([node, func]),
	loading: bool,
	outline: bool,
	size: oneOf(['small', 'default', 'large']),
	text: bool,
	type: oneOf(['default', 'white', 'primary', 'danger', 'success']),
};

ButtonWrapper.defaultProps = {
	type: 'default',
};

export { Button as EmotionButton };
export default ButtonWrapper;
