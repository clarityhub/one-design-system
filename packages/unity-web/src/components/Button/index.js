export { default, EmotionButton } from './Button';