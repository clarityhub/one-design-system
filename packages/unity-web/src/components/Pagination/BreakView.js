import React from 'react';
import { string, node, oneOfType } from 'prop-types';

const BreakView = props => {
	const { breakLabel } = props;

	return (
		<li>
			{breakLabel}
		</li>
	);
};

BreakView.propTypes = {
	breakLabel: oneOfType([string, node]),
};

export default BreakView;