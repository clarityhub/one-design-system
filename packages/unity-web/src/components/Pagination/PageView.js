import React from 'react';
import { string, func, number, bool } from 'prop-types';

const PageView = ({ pageRenderer, ...props }) => {
	let pageClassName = props.pageClassName;
	let pageLinkClassName = props.pageLinkClassName;
	let page = props.page;

	const onClick = props.onClick;
	const href = props.href;

	let ariaLabel =
    'Page ' +
    props.page +
    (props.extraAriaContext ? ' ' + props.extraAriaContext : '');
	let ariaCurrent = null;

	if (props.selected) {
		ariaCurrent = 'page';
		ariaLabel = 'Page ' + props.page + ' is your current page';

		if (typeof pageClassName !== 'undefined') {
			pageClassName = pageClassName + ' ' + props.activeClassName;
		} else {
			pageClassName = props.activeClassName;
		}

		if (typeof pageLinkClassName !== 'undefined') {
			if (typeof props.activeLinkClassName !== 'undefined') {
				pageLinkClassName = pageLinkClassName + ' ' + props.activeLinkClassName;
			}
		} else {
			pageLinkClassName = props.activeLinkClassName;
		}
	}

	return (
		<li className={pageClassName}>
			{pageRenderer({
				onClick,
				className: pageLinkClassName,
				href,
				ariaLabel,
				ariaCurrent,
				onKeyPress: onClick,
				page,
			})}
		</li>
	);
};

PageView.propTypes = {
	activeClassName: string,
	activeLinkClassName: string,
	extraAriaContext: string,
	onClick: func.isRequired,
	page: number.isRequired,
	pageClassName: string,
	pageLinkClassName: string,
	pageRenderer: func.isRequired,
	selected: bool.isRequired,
};

export default PageView;