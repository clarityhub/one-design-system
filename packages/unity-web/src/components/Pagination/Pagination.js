import React from 'react';
import { number, func } from 'prop-types';
import isPropValid from '@emotion/is-prop-valid';
import { css } from '@emotion/core';
import styled from '@emotion/styled';
import colors from '@clarityhub/unity-core/lib/colors';

import breakpoints from '../../theme/breakpoints';
import Link from '../Link';
import Paginate from './PaginateBoxView';

// XXX need a Link renderer (with on clicks and stuff)

const Container = styled.div`
    margin: 3rem 1rem;
    text-align: center;

    li {
        display: inline-block;

    }

    li + li {
        margin-left: 1rem;
    }

    @media (max-width: ${breakpoints.smallMaxWidth}px) {
        li:first-of-type,
        li:last-of-type {
            display: block;
            margin: 1rem;
        }
    }
`;

const Item = styled('div', {
	shouldForwardProp: prop => isPropValid(prop) && !['selected'].includes(prop),
})`
	a {
        border: 1px solid ${colors.gray.default};
		border-radius: 50%;
        color: ${colors.primary.default};
		font-size: 0.9rem;
        text-align: center;
		height: 2.25rem;
		width: 2.25rem;
		display: flex;
		align-items: center;
		justify-content: center;

        &:hover {
			background-color: ${colors.muted.default};
			border: 1px solid ${colors.gray.default};
        }

        &:visited {
			border: 1px solid ${colors.gray.default};
            color: ${colors.primary.default};
        }

        ${({ selected }) => selected && css`
            background-color: ${colors.primary.default};
			color: ${colors.white.default} !important;
			border: 1px solid ${colors.primary.default};
			box-shadow: ${colors.shadow.default};

			&:hover {
				background-color: ${colors.primary.default};
				border: 1px solid ${colors.primary.default};
			}

            &:visited {
                color: ${colors.white.default} !important;
            }
        `}
    }
`;

const Pagination = ({
	index,
	pageCount,
	nextRenderer = ({ index, disabled }) => {
		const nextUrl = (index + 1).toString();

		return (
			<Link href={`/${nextUrl}`}>
				&gt;
			</Link>
		);
	},
	previousRenderer = ({ index, disabled }) => {
		const previousUrl = index - 1 === 1 ? '' : (index - 1).toString();

		return (
			<Link href={`/${previousUrl}`}>
                &lt;
			</Link>
		);
	},
	pageRenderer = ({ page }) => {
		const url = page === 1 ? '' : page;

		return (
			<Link href={`/${url}`}>
				{page}
			</Link>
		);
	},
}) => {
	if (pageCount === 1) {
		return null;
	}

	return (
		<Container>
			<Paginate
				pageCount={pageCount}
				pageRangeDisplayed={3}
				marginPagesDisplayed={1}
				selected={index - 1}
				nextRenderer={({ disabled }) => (
					disabled ? null : (
						<Item>
							{nextRenderer({ index, disabled })}
						</Item>
					)
				)}
				previousRenderer={({ disabled }) => (
					disabled ? null : (
						<Item>
							{previousRenderer({ index, disabled })}
						</Item>
					)
				)}
				pageRenderer={({ page }) => {
					return (
						<Item selected={String(page) === String(index)}>
							{pageRenderer({ page })}
						</Item>
					);
				}}
			/>
		</Container>
	);
};

Pagination.propTypes = {
	first: number,
	index: number.isRequired,
	last: number,
	nextRenderer: func,
	pageCount: number.isRequired,
	pageRenderer: func,
	previousRenderer: func,
};

export default Pagination;
