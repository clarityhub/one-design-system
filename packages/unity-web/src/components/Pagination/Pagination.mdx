---
menu: Components
name: Pagination
---

import { Playground, PropsTable } from 'docz';
import Pagination from './Pagination';
import Link from '../Link';

# Pagination

## \<Pagination /\> component

```js
import Pagination from '@clarityhub/unity-web/lib/components/Pagination';
```

## Prop Types

<PropsTable of={Pagination} />

## Examples

<Playground>
   <Pagination
        index={4}
        first={1}
        last={120}
        pageCount={120}
    />

    <Pagination
        index={1}
        first={1}
        last={120}
        pageCount={120}
    />
</Playground>

### Override the link renderers

<Playground>
    <Pagination
        index={1}
        pageCount={120}

        nextRenderer={({ index, disabled }) => {
            const nextUrl = (index + 1).toString();

            return (
                <Link href={`#/${nextUrl}`}>
                    Next &raquo;
                </Link>
            );
        }}
        previousRenderer={({ index, disabled }) => {
            const previousUrl = index - 1 === 1 ? '' : (index - 1).toString();

            return (
                <Link href={`#/${previousUrl}`}>
                    &laquo; Previous
                </Link>
            );
        }}
        pageRenderer={({ page }) => {
            const url = page === 1 ? '' : page;

            return (
                <Link href={`#/${url}`}>
                    {page}
                </Link>
            );
        }}
    />
</Playground>