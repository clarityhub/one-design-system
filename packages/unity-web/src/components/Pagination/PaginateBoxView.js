
import React, { Component } from 'react';
import { bool, func, node, number, oneOfType, string } from 'prop-types';
import PageView from './PageView';
import BreakView from './BreakView';

export default class PaginationBoxView extends Component {
  static propTypes = {
  	activeClassName: string,
  	activeLinkClassName: string,
  	breakClassName: string,
  	breakLabel: oneOfType([string, node]),
  	breakLinkClassName: string,
  	containerClassName: string,
  	disableInitialCallback: bool,
  	disabledClassName: string,
  	extraAriaContext: string,
  	marginPagesDisplayed: number,
  	nextClassName: string,
  	nextLabel: node,
  	nextLinkClassName: string,
  	nextRenderer: func,
  	onPageChange: func,
  	pageClassName: string,
  	pageCount: number,
  	pageLinkClassName: string,
  	pageRangeDisplayed: number,
  	pageRenderer: func,
  	previousClassName: string,
  	previousLabel: node,
  	previousLinkClassName: string,
  	previousRenderer: func,
  	selected: number,
  };

  static defaultProps = {
  	pageCount: 10,
  	pageRangeDisplayed: 2,
  	marginPagesDisplayed: 3,
  	activeClassName: 'selected',
  	previousClassName: 'previous',
  	nextClassName: 'next',
  	previousLabel: 'Previous',
  	nextLabel: 'Next',
  	breakLabel: '...',
  	disabledClassName: 'disabled',
  	disableInitialCallback: false,
  };

  componentDidMount() {
  	const { initialPage, disableInitialCallback } = this.props;
  	// Call the callback with the initialPage item:
  	if (typeof initialPage !== 'undefined' && !disableInitialCallback) {
  		this.callCallback(initialPage);
  	}
  }

  handlePreviousPage = evt => {
  	const { selected } = this.props;
  	evt.preventDefault ? evt.preventDefault() : (evt.returnValue = false);
  	if (selected > 0) {
  		this.handlePageSelected(selected - 1, evt);
  	}
  };

  handleNextPage = evt => {
  	const { selected, pageCount } = this.props;

  	evt.preventDefault ? evt.preventDefault() : (evt.returnValue = false);
  	if (selected < pageCount - 1) {
  		this.handlePageSelected(selected + 1, evt);
  	}
  };

  handlePageSelected = (selected, evt) => {
  	evt.preventDefault ? evt.preventDefault() : (evt.returnValue = false);

  	if (this.props.selected === selected) return;

  	this.setState({ selected: selected });

  	// Call the callback with the new selected item:
  	this.callCallback(selected);
  };

  getForwardJump() {
  	const { selected, pageCount, pageRangeDisplayed } = this.props;

  	const forwardJump = selected + pageRangeDisplayed;
  	return forwardJump >= pageCount ? pageCount - 1 : forwardJump;
  }

  getBackwardJump() {
  	const { selected, pageRangeDisplayed } = this.props;

  	const backwardJump = selected - pageRangeDisplayed;
  	return backwardJump < 0 ? 0 : backwardJump;
  }

  handleBreakClick = (index, evt) => {
  	evt.preventDefault ? evt.preventDefault() : (evt.returnValue = false);

  	const { selected } = this.props;

  	this.handlePageSelected(
  		selected < index ? this.getForwardJump() : this.getBackwardJump(),
  		evt
  	);
  };

  callCallback = selectedItem => {
  	if (
  		typeof this.props.onPageChange !== 'undefined' &&
      typeof this.props.onPageChange === 'function'
  	) {
  		this.props.onPageChange({ selected: selectedItem });
  	}
  };

    pageRenderer = (props) => {
    	const { pageRenderer } = this.props;
    	const {
    		page,
    		onClick,
    		className,
    		ariaLabel,
    		ariaCurrent,
    	} = props;

    	if (pageRenderer) {
    		return pageRenderer(props);
    	}

    	return (
    		<a
    			onClick={onClick}
    			role="button"
    			className={className}
    			href="/#"
    			tabIndex="0"
    			aria-label={ariaLabel}
    			aria-current={ariaCurrent}
    			onKeyPress={onClick}
    		>
    			{page}
    		</a>
    	);
    }

    previousRenderer = (props) => {
    	const { previousRenderer } = this.props;
    	const {
    		onClick,
    		className,
    		onKeyPress,
    		ariaDisabled,
    		children,
    	} = props;

    	if (previousRenderer) {
    		return previousRenderer(props);
    	}

    	return (
    		<a
    			onClick={onClick}
    			className={className}
    			tabIndex="0"
    			href="/#"
    			role="button"
    			onKeyPress={onKeyPress}
    			aria-disabled={ariaDisabled}
    		>
    			{children}
    		</a>
    	);
    }

    nextRenderer = (props) => {
    	const { nextRenderer } = this.props;
    	const {
    		onClick,
    		className,
    		onKeyPress,
    		ariaDisabled,
    		children,
    	} = props;

    	if (nextRenderer) {
    		return nextRenderer(props);
    	}

    	return (
    		<a
    			onClick={onClick}
    			className={className}
    			tabIndex="0"
    			href="/#"
    			role="button"
    			onKeyPress={onKeyPress}
    			aria-disabled={ariaDisabled}
    		>
    			{children}
    		</a>
    	);
    }

    getPageElement(index) {
    	const {
    		pageClassName,
    		pageLinkClassName,
    		activeClassName,
    		activeLinkClassName,
    		extraAriaContext,
    		selected,
    	} = this.props;

    	return (
    		<PageView
    			key={index}
    			onClick={this.handlePageSelected.bind(null, index)}
    			selected={selected === index}
    			pageClassName={pageClassName}
    			pageLinkClassName={pageLinkClassName}
    			activeClassName={activeClassName}
    			activeLinkClassName={activeLinkClassName}
    			extraAriaContext={extraAriaContext}
    			pageRenderer={this.pageRenderer}
    			page={index + 1}
    		/>
    	);
    }

  pagination = () => {
  	const items = [];
  	const {
  		pageRangeDisplayed,
  		pageCount,
  		marginPagesDisplayed,
  		breakLabel,
  		breakClassName,
  		breakLinkClassName,
  		selected,
  	} = this.props;

  	if (pageCount <= pageRangeDisplayed) {
  		for (let index = 0; index < pageCount; index++) {
  			items.push(this.getPageElement(index));
  		}
  	} else {
  		let leftSide = pageRangeDisplayed / 2;
  		let rightSide = pageRangeDisplayed - leftSide;

  		// If the selected page index is on the default right side of the pagination,
  		// we consider that the new right side is made up of it (= only one break element).
  		// If the selected page index is on the default left side of the pagination,
  		// we consider that the new left side is made up of it (= only one break element).
  		if (selected > pageCount - pageRangeDisplayed / 2) {
  			rightSide = pageCount - selected;
  			leftSide = pageRangeDisplayed - rightSide;
  		} else if (selected < pageRangeDisplayed / 2) {
  			leftSide = selected;
  			rightSide = pageRangeDisplayed - leftSide;
  		}

  		let index;
  		let page;
  		let breakView;
  		let createPageView = index => this.getPageElement(index);

  		for (index = 0; index < pageCount; index++) {
  			page = index + 1;

  			// If the page index is lower than the margin defined,
  			// the page has to be displayed on the left side of
  			// the pagination.
  			if (page <= marginPagesDisplayed) {
  				items.push(createPageView(index));
  				continue;
  			}

  			// If the page index is greater than the page count
  			// minus the margin defined, the page has to be
  			// displayed on the right side of the pagination.
  			if (page > pageCount - marginPagesDisplayed) {
  				items.push(createPageView(index));
  				continue;
  			}

  			// If the page index is near the selected page index
  			// and inside the defined range (pageRangeDisplayed)
  			// we have to display it (it will create the center
  			// part of the pagination).
  			if (index >= selected - leftSide && index <= selected + rightSide) {
  				items.push(createPageView(index));
  				continue;
  			}

  			// If the page index doesn't meet any of the conditions above,
  			// we check if the last item of the current "items" array
  			// is a break element. If not, we add a break element, else,
  			// we do nothing (because we don't want to display the page).
  			if (breakLabel && items[items.length - 1] !== breakView) {
  				breakView = (
  					<BreakView
  						key={index}
  						breakLabel={breakLabel}
  						breakClassName={breakClassName}
  						breakLinkClassName={breakLinkClassName}
  						onClick={this.handleBreakClick.bind(null, index)}
  					/>
  				);
  				items.push(breakView);
  			}
  		}
  	}

  	return items;
  };
  

  render() {
  	const {
  		disabledClassName,
  		previousClassName,
  		nextClassName,
  		pageCount,
  		containerClassName,
  		previousLinkClassName,
  		previousLabel,
  		nextLinkClassName,
  		nextLabel,
  		selected,
  	} = this.props;

  	const previousClasses =
      previousClassName + (selected === 0 ? ` ${disabledClassName}` : '');
  	const nextClasses =
      nextClassName +
      (selected === pageCount - 1 ? ` ${disabledClassName}` : '');

  	const previousAriaDisabled = selected === 0;
  	const nextAriaDisabled = selected === pageCount - 1;

  	return (
  		<ul className={containerClassName}>
  			<li className={previousClasses}>
  				{this.previousRenderer({
  					onClick: this.handlePreviousPage,
  					className: previousLinkClassName,
  					onKeyPress: this.handlePreviousPage,
  					disabled: previousAriaDisabled,
  					ariaDisabled: String(previousAriaDisabled),
  					children: previousLabel,
  				})}
  			</li>

  			{this.pagination()}

  			<li className={nextClasses}>
  				{this.nextRenderer({
  					onClick: this.handleNextPage,
  					className: nextLinkClassName,
  					onKeyPress: this.handleNextPage,
  					disabled: nextAriaDisabled,
  					ariaDisabled: String(nextAriaDisabled),
  					children: nextLabel,
  				})}
  			</li>
  		</ul>
  	);
  }
}