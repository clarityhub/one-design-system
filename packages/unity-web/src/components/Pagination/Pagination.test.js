import React from 'react';
import renderer from 'react-test-renderer';

import Link from '../Link';
import Button from '../Button';

import Pagination from './Pagination';

describe('<Pagination />', () => {
	it('renders paginiation bar', () => {
		const node = (
			<Pagination
				index={4}
				first={1}
				last={120}
				pageCount={120}
			/>
		);
		const tree = renderer.create(node).toJSON();

		expect(tree).toMatchSnapshot();
	});

	it('uses renderers', () => {
		const node = (
			<Pagination
				index={1}
				pageCount={120}

				nextRenderer={({ index, disabled }) => {
					return (
						<Button>
                            Next &raquo;
						</Button>
					);
				}}
				previousRenderer={({ index, disabled }) => {
					return (
						<Button>
                            &laquo; Previous
						</Button>
					);
				}}
				pageRenderer={({ page }) => {
					const url = page === 1 ? '' : page;

					return (
						<Link href={`#/${url}`}>
							{page}
						</Link>
					);
				}}
			/>
		);
		const tree = renderer.create(node).toJSON();

		expect(tree).toMatchSnapshot();
	});
});