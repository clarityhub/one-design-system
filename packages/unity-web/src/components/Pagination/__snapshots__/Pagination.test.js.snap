// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> renders paginiation bar 1`] = `
<div
  className="css-13p1w26-Container e171o4c40"
>
  <ul>
    <li
      className="previous"
    >
      <div
        className="css-repihg-Item e171o4c41"
      >
        <a
          className="css-i8x1an"
          href="/3"
        >
          &lt;
        </a>
      </div>
    </li>
    <li>
      <div
        className="css-repihg-Item e171o4c41"
      >
        <a
          className="css-i8x1an"
          href="/"
        >
          1
        </a>
      </div>
    </li>
    <li>
      ...
    </li>
    <li>
      <div
        className="css-repihg-Item e171o4c41"
      >
        <a
          className="css-i8x1an"
          href="/3"
        >
          3
        </a>
      </div>
    </li>
    <li
      className="selected"
    >
      <div
        className="css-1fxzov6-Item-Item e171o4c41"
      >
        <a
          className="css-i8x1an"
          href="/4"
        >
          4
        </a>
      </div>
    </li>
    <li>
      <div
        className="css-repihg-Item e171o4c41"
      >
        <a
          className="css-i8x1an"
          href="/5"
        >
          5
        </a>
      </div>
    </li>
    <li>
      ...
    </li>
    <li>
      <div
        className="css-repihg-Item e171o4c41"
      >
        <a
          className="css-i8x1an"
          href="/120"
        >
          120
        </a>
      </div>
    </li>
    <li
      className="next"
    >
      <div
        className="css-repihg-Item e171o4c41"
      >
        <a
          className="css-i8x1an"
          href="/5"
        >
          &gt;
        </a>
      </div>
    </li>
  </ul>
</div>
`;

exports[`<Pagination /> uses renderers 1`] = `
<div
  className="css-13p1w26-Container e171o4c40"
>
  <ul>
    <li
      className="previous disabled"
    />
    <li
      className="selected"
    >
      <div
        className="css-1fxzov6-Item-Item e171o4c41"
      >
        <a
          className="css-i8x1an"
          href="#/"
        >
          1
        </a>
      </div>
    </li>
    <li>
      <div
        className="css-repihg-Item e171o4c41"
      >
        <a
          className="css-i8x1an"
          href="#/2"
        >
          2
        </a>
      </div>
    </li>
    <li>
      <div
        className="css-repihg-Item e171o4c41"
      >
        <a
          className="css-i8x1an"
          href="#/3"
        >
          3
        </a>
      </div>
    </li>
    <li>
      <div
        className="css-repihg-Item e171o4c41"
      >
        <a
          className="css-i8x1an"
          href="#/4"
        >
          4
        </a>
      </div>
    </li>
    <li>
      ...
    </li>
    <li>
      <div
        className="css-repihg-Item e171o4c41"
      >
        <a
          className="css-i8x1an"
          href="#/120"
        >
          120
        </a>
      </div>
    </li>
    <li
      className="next"
    >
      <div
        className="css-repihg-Item e171o4c41"
      >
        <button
          className="-default css-1qfeqqs-Button-Button entsyi60"
          type="button"
        >
          Next Â»
        </button>
      </div>
    </li>
  </ul>
</div>
`;
